### DISCUSSION TITLE
Is it safe to migrate to .NET 8.0

### DISCUSSION THREAD

**Discussion by williambuchanan2** [dotnet/maui#15326] [General]
Hi all

From reading I have found that the Maui team are fixing issues in .NET 8, and then if the issue is critical enough they are then backfixing it in .NET 7. I believe also some issues are not getting fixed in .NET 7 (i.e. they are only being fixed in .NET 8, so .NET 7 is technically already abandoned).

First, is this true?

If so, would it be wise to migrate everything to .NET 8 now?

I need to have this app released well before the November release date of .NET 8, so it would be a gamble to migrate, but with the current version being such a disaster would there be more chance of getting a stable release from a pre-release version of .NET 8 than sticking to .NET 7?

**Comment by Vroomer** [comment]
Is it safe? No. New PRs tend to break old things and the .NET 8 branch is definitely not stable. For example there is a big issue with BindingContext on Windows that I recently reported. However there is a good chance that the .NET 8 will be much better in release candidates than the current .NET 7 releases, so if you are developing it now and you are blocked by many bugs in .NET 7, I would give it a try.

Personally I find the nightly builds of main branch even better for development. You can find them here: https://dev.azure.com/xamarin/public/_build?definitionId=57 They are built against .NET 7, in my experience it's quite stable on Windows and it has the latest bug fixes. I hope that by the time I finish the app, there will be stable release out already.

**Reply by williambuchanan2** [reply]
That's pretty much what I was thinking... How do you work with those nightly builds? Are you downloading the artifacts? I tried but wasn't working...

**Reply by Vroomer** [reply]
Download the artifact, put it into a folder that is shared with Nuget Manager and put this into .csproj:
`<MauiVersion>7.0.0-ci.net7.17908</MauiVersion>`

Usually it throws some errors regarding splash screen resources etc. I found that removing those parts from .csproj, saving it, insertng it back and saving it again makes it possible to succesfuly rebuild the project. You might also have to empty bin/obj folders. I didn't try iOS though.

**Comment by giuseppenovielli** [comment]
@williambuchanan2 Until .NET8 if MAUI is not good for you, another choice, only if you can, try to build with Xamarin.Forms that is stable.

**Reply by williambuchanan2** [reply]
We had originally planned to do this project in Xamarin but decided to hold off once they announced Maui. After years of delay here we are - still working on it. The problem with going with Xamarin is that they are killing it off next year, so that's not a good long term plan I think.

**Comment by Xyncgas** [comment]
.NET 8 should be selectable as new target framework on November in visual studio

**Reply by williambuchanan2** [reply]
I tried it in preview but it doesn't build. Might leave it a few weeks and try again, but we are using azure pipelines so a bit concerned about how it will work with that - it's always a bit unreliable for a while...

**Comment by nor0x** [comment]
simple answer... NO! 
nothing in maui is safe right now 😂

**Reply by williambuchanan2** [reply]
Aint that thr truth...

**Comment by Stedy59** [comment]
Yes!!! You can using Visual Studio 17.7.0 Preview 1.0

I have converted one app, and started 2 new applications! [.NET8 Preview is available in the template dropdown!]

Make sure you install .net8 using Visual Studio installer, and then, install the latest MAUI as detailed...
**`dotnet workload install maui`**

One thing I have definitely noticed is the app runtime speed has improved... ALOT! I am currently working on a custom framework that includes over 20 custom controls... ie: TabView, AnimatedNavigationPage, SlideViews, Dropdown, Custom DialogView (popup/messagebox). etc.

Here is a link to a video that show SOME of the functionality ... (in progress) [ https://youtu.be/S_-Y2zZhuRs]( https://youtu.be/S_-Y2zZhuRs)
If you have any question, please feel free to PM me...

**Reply by williambuchanan2** [reply]
Ok that's good to hear... Are you using Azure pipelines as well? We need to deploy the app to people while it's under development so would need to make sure our pipelines work as well.

**Reply by Stedy59** [reply]
Unfortunately, no...

**Comment by tylerwilson** [comment]
My issue is that I have a ,NET 7 MAUI app, and I get crashes in the real world, and then I see they are fixed in MAUI 8 preview 1 or 8 preview 4, etc. I am using Visual Studio for Mac, which does not properly support .NET/MAUI 8 either.

It seems to me it was a bad call to tie major MAUI releases with .NET releases. We can't wait a year to do proper updates. There is this whole NuGet system in place, let's use that. 

I understand there is some way to use updates via NuGet (though I cannot find the instructions at the moment) but if memory serves it is a very manual process. Why are they not available on a MS NuGet server that we can just add to our IDEs

**Reply by williambuchanan2** [reply]
Yes that is a good point. I have been using VS for Mac as well (which is as bug ridden as everything else)...

**Comment by nicjay** [comment]
Please do not use a pre-release version in production, wait until RC1 at minimum when MS typically gives their "go-live" approval. .NET 7 is still getting service releases.

That said, I agree a year is way too long to wait for major fixes not being backported. This package has been particularly useful in the meantime: https://www.nuget.org/packages/PureWeen.Maui.FixesAndWorkarounds

**Comment by Xyncgas** [comment]
Would be awesome our software is built on block chain, all future update happens on the block chain, no windows update required no patch delivery required, people just build the newest software starting from the genesis block like we do with bitcoin currently.

**Reply by Xyncgas** [reply]
if you are using such technology, you can update your dependency, change from .NET 7 to .NET 8, doesn't matter what happens on the developer's machine, it's magically delivered and appeared on consumer's machine, there are different ways to do this.

.NET itself is based on block chain, the entire code base is compiled to IL for JIT which is what's happening currently but now we keep track of changes happens on the compiled code on the block chain, this record on the block chain is merged with any additional code from user's project, ending up a block chain that records the changes for the entire project. When people use our application, they fetch the genesis block which could've been what our project looks like on day one using .NET 6, then as they fetch newer block they are essentially walking through the life of our product, fetching update after update, such as what our project looks like last year using .NET 7, now .NET 8, in a incremental building manner.

Or we can implement this directly in our application today by building a virtual machine and only ship virtual instructions, which we can put on a server or in the distributed computers that participate in a private update network, we can do this only for the critical business rules or the entire application

**Comment by s-beji** [comment]
I have tried the .net 8 preview and I strongly recommend not using it for production as it is less stable than .net 7 and also introduces new bugs so there is no real reason to use it now for anything other than testing and reporting, but never for production or anything close to it

**Reply by AndreasReitberger** [reply]
The problem is, that you neither can use `net7` for production. At least not for `iOS` with the `Keyboard` overlapping the `Entry` controls. I would love and wait for the next SR of `net7`.

**Reply by s-beji** [reply]
thats a whole diffrent topic if maui .net 7 is really production ready (my opinion its not for complex project sutited at least for now) but i am sure there is an workaround for the keyboard issue atleast there were for android since i dont own an apple device i cant help you with that issue

**Reply by williambuchanan2** [reply]
Exactly right - .NET 6 and .NET 7 are not prod ready - so there is actually no solution to the problem at the moment...

**Reply by UkeHa** [reply]
@AndreasReitberger you can use this fix for ios to not overlap anymore - [fix](https://github.com/massijay/iOSMauiScrollUpOnKeyboardShow/blob/main/Platforms/iOS/Utils/Extensions.cs)

**Reply by jbtdevgit** [reply]
Nope, even .net 7 for iOS. There is tantamount of bugs when it comes to ipas and submissions, crashes, even has troubles with icons. You can't even rebuild on release mode in Windows for iOS to create a release. Such a bummer, I never had this problem back in Xamarin.Forms.

**Comment by gabsamples6** [comment]
@UkeHa  I am looking for a fix for the keyboard overlap and that does not work for me - it just works in some scenarios...
Has anyone implemented this fix for keyboard overlap in a chat page kind of look and feel with the entry at the bottom like whats'up... 

Appreciate any feedback on this

**Reply by williambuchanan2** [reply]
I did something similar. I ended up using the IQKeyboardManager nuget package. It's ok but still not perfect so I am still looking for a good solution.

**Reply by UkeHa** [reply]
@gabsamples6, using devexpress keyboards + the previously mentioned fix pretty much solved it for me.

**Comment by thomasgalliker** [comment]
Ok guys. I migrated a pretty large app from XF to .NET 7 MAUI and I ended up in a desaster. We talk about 200h+ of migration effort until I had to realize, there is no chance .NET 7 MAUI will ever fix all the remaining issues the migration produced (rendering issues, slow navigation, NREs, … a lot of troubles).

Desperately, I tried the latest preview of .NET 8 MAUI which did not seem to solve any of the (obvious) issues.

**Reply by williambuchanan2** [reply]
It's almost impossible to make a maui app built from the ground up work - I would imagine the chance of a successful migration would be very close to zero.

**Reply by UkeHa** [reply]
I have to disagree, with the most recent releases and all the workarounds it has become stable enough to release something. I've actually migrated an app from xamarin. It took way too long, things were completely broken or made us use devexpress components to fix issues on either android or iOS, but now it works. With .net8 it should become stable enough to really use it. The visual studio integration has also come a long way.

**Reply by williambuchanan2** [reply]
There are still bugs in .NET 7 which are preventing us from releasing (one example - if you have images in your app which you load dynamically it crashes Android). So, yes you can get something which looks like a working app, but it is far from stable without a lot of workarounds. Layout is impossible to get nice due to it not working consistently across platforms, so again a bunch of workarounds for that.

I have tried .NET 8 a few times - even the example click counter doesn't work on ios.

So I would say it is still a long way from being a stable and easy to use platform, and certainly not something you should entertain unless you have a lot of time to burn.

**Reply by thomasgalliker** [reply]
It feels like we're back in 2015, when Xamarin wasn't part of Microsoft and the struggle was real. Let's keep finger's crossed and hope Microsoft will sort it out.

**Comment by giuseppenovielli** [comment]
https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-rc-2/#bonus-2-xcode-15-and-android-api-34

Xamarin.Forms can build with iOS 17, Android 14, if you don't need latest platform API.
